Make a how to for creating a Sorting (#2944)
SpikeInterface/spikeinterface: doc/how_to/make_a_sorting.rst

    # in this case we are making a monosegment sorting
    my_sorting = NumpySorting.from_times_labels(times_list = [1,2,3,4],
                                                labels_list = [0,1,0,1],
                                                sampling_frequency = 30_000.0
                                                )

With a unit dictionary
^^^^^^^^^^^^^^^^^^^^^^

We can also use a dictionary where each unit is a key and its spike times are values.
This is entered as either a list of dicts with each dict being a segment or as a single
dict for monosegment. We still need to separately specify the sampling_frequency

.. code-block:: python

    from spikeinterface.core import NumpySorting

    my_sorting = NumpySorting.from_unit_dict(units_dict_list={'0': [1,3],
                                                              '1': [2,4]
                                                              },
                                            sampling_frequency=30_000.0
                                           )

With Neo SpikeTrains
^^^^^^^^^^^^^^^^^^^^

Finally since SpikeInterface is tightly integrated with the Neo project you can create

> a sorting from :code:`Neo.SpikeTrain` objects. See Neo documentation for more information on

Author: I forgot how to do the auto-link so if someone can just remind me in review we should link to the Neo docs.

Reviewer: Works like:
Please read the :doc:`Neo documentation<neo:index>`.

(currently we've got neo and probeinterface registered to be used like this)

Make your own Sorting
=====================

> Why make a :code:`Sorting`?

Author: I'm bad at that for RTD. So if you have an idea definitely suggest. Linking is the bane of my existence, but I like it!

The :code:`Sorting` object is one of the core objects within the SpikeInterface library
along with :code:`Recording` and :code:`SortingAnalyzer`. Although SpikeInterface has
many tools for reading sorting formats you may have some data in a nonstandard format
(e.g. old csv file). If this is the case you would need to make your own :code:`Sorting`.

At a fundamental level the :code:`Sorting` is a series of spike times and a series of
labels for each spike along with some associated metadata. Thus by providing this
information you can easily make a :code:`Sorting` object to be used for various analyses
(e.g. correlograms or for generating a :code:`SortingAnalyzer`)

Making a :code:`Sorting`
------------------------

For most formats the :code:`Sorting` is automatically generated. For example one could do

.. code-block:: python

    from spikeinterface.extractors import read_kilosort, read_phy

    # For kilosort/phy files we can use either reader
    ks_sorting = read_kilosort('path/to/folder')

>     phy_sorting = read_phy('path/to/folder')

Author: Fair. I prefer focused.

This :code:`Sorting` contains important information about your spike trains including
the spike times (i.e. when the neurons were actually firing) the unit labels (i.e.
who the spikes belong to. Also called cluster ids by some sorters), the unit ids (the unique
set of unit labels) and the sampling_frequency. To make your own :code:`Sorting` object you can

> use :code:`NumpySorting`. It is important to note if you are new to spike trains that they

Reviewer 2: ```suggestion
use :code:`NumpySorting`. It is important to note that in SpikeiInterface spike trains are handled internally in samples/frames rather than in seconds and we use the sampling frequency to ...
```

I think something along these lines is more precise. What do you think?

are typically stored in samples/frames rather than in seconds. So you should input the times
in samples/frames. The sampling_frequency allows for easily switching between samples and seconds.

> There are 3 options (along with making a NumpySorting from another sorting which will not be covered here):

Reviewer 2: I don't think the detail in the parenthesis is needed in a how to. I feel we don't need to be thoroug.

> There are 3 options (along with making a NumpySorting from another sorting which will not be covered here):

Author: This was the discussion Heberto and I had above. I would argue the field is used to the spike times term so I think we say your spike times in samples to be clearer.

With lists of spike trains and spike labels
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> In this case we need a list or array (or lists of lists for multisegment) of spike times,

Reviewer 2: ```suggestion
In this case we need a list or array (or lists of lists for multisegment) of spike frames,
```

I think that `times_list` is a bad argument name as it is prone to confusion.

Author: I agree. I would think we should say 

spike times (in frames) or spike times in frames instead. Saying spike times is common in the field to mean in frames or in seconds so I think it would be more standard in the field to say spike times and then specify the units. What do you think?

Reviewer 2: I think that you are in a better position to make that call than me.  I trust you in this one.

unit labels, sampling_frequency and optional unit_ids if you want specific labels to be
used (in this case we only create the :code:`Sorting` based on the requested unit_ids).

.. code-block:: python

    from spikeinterface.core import NumpySorting

    # in this case we are making a monosegment sorting

>     my_sorting = NumpySorting.from_times_labels(times_list = [1,2,3,4],

Author: That's fair. I was being lazy :)

                                                labels_list = [0,1,0,1],

>                                                 sampling_frequency = 30_000.0

Author: Yes. It is. You can put underscores into any number for spacing :)

>                                                 sampling_frequency = 30_000.0

Reviewer 3: 🤯 cool!!

>                                                 sampling_frequency = 30_000.0

Reviewer 2: It is very nice, man!

>                                                 )

Author: For np.arrays I agree. That's fine.

I was purposely avoiding the multisegment because people said that is confusing for people working with mono segments. But I can add. Maybe one example lower down? That way we have an easy example and people can shut off their brains if they don't need the multisegment?

>                                                 )

Reviewer 4: OK for the mono segment without list but this must be explicit in the paragraph somewhere.

>                                                 )

Author: Yep will do. My plan is to incorporate all changes on Friday when I can sit and do it carefully :)

>                                                 )

Reviewer 2: I also believe that in general is better to explain things in the mono segment and then we can mention the multi segment after. Is a complexity that is probably not necessary for most cases and concepts.

With a unit dictionary
^^^^^^^^^^^^^^^^^^^^^^

> We can also use a dictionary where each unit is a key and its spike times are values.

Reviewer 2: ```suggestion
We can also use a dictionary where each unit is a key and a list of spike frames are passed as values.
```
In the same spirit of time vs frames distinction above.

Author: See comment above and then we can decide!

This is entered as either a list of dicts with each dict being a segment or as a single
dict for monosegment. We still need to separately specify the sampling_frequency

.. code-block:: python

    from spikeinterface.core import NumpySorting

    my_sorting = NumpySorting.from_unit_dict(units_dict_list={'0': [1,3],
                                                              '1': [2,4]

>                                                               },

Reviewer 4: here also we should show an example with multi segment in mind.

                                             sampling_frequency=30_000.0
                                           )

With Neo SpikeTrains
^^^^^^^^^^^^^^^^^^^^

Finally since SpikeInterface is tightly integrated with the Neo project you can create
a sorting from :code:`Neo.SpikeTrain` objects. See Neo documentation for more information on
using :code:`Neo.SpikeTrain`'s.

.. code-block:: python

    from spikeinterface.core import NumpySorting

    # neo_spiketrain is a Neo spiketrain object
    my_sorting = NumpySorting.from_neo_spiketrain_list(neo_spiketrain,

>                                                        sampling_frequency=30_000.0)

Reviewer: I'd add a line, just to wrap up the guide. Something like:

"
Now that you've created a Sorting object, you can combine it with a recording to make a :ref:`Sorting Analyzer <sphx_glr_tutorials_core_plot_4_sorting_analyzer.py>`, or start visualising by using the :py:func:`~spikeinterface.widgets.plot_crosscorrelograms` function.
"

But up to you!
